external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala

```diff
@@ -0,0 +1,199 @@
+      offset match {
+        case o: KafkaSourceOffset => o
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Invalid conversion from offset of ${offset.getClass} to $getClass")
+      }
+    }
+  }
+
+  /**
+   * Returns [[KafkaSourceOffset]] from a variable sequence of (topic, partitionId, offset)
+   * tuples.
+   */
+  def apply(data: (String, Int, Long)*): KafkaSourceOffset = {
+    val map = data.map { case (topic, partition, offset) =>
+        TopicAndPartition(topic, partition) -> offset }.toMap
+    KafkaSourceOffset(map)
+  }
+}
+
+
+/** A [[Source]] that reads data from Kafka */
+private[kafka] case class KafkaSource(
+    topics: Set[String],
+    params: Map[String, String])(implicit sqlContext: SQLContext) extends Source with Logging {
+
+  type OffsetMap = Map[TopicAndPartition, Long]
+
+  implicit private val encoder = ExpressionEncoder.tuple(
+    ExpressionEncoder[Array[Byte]](), ExpressionEncoder[Array[Byte]]())
+
+  @transient private val logicalPlan = StreamingRelation(this)
+  @transient private val kc = new KafkaCluster(params)
+  @transient private val topicAndPartitions = KafkaCluster.checkErrors(kc.getPartitions(topics))
+  @transient private[kafka] val initialOffsets = getInitialOffsets()
+
+  override def schema: StructType = encoder.schema
+
+  /** Returns the next batch of data that is available after `start`, if any is available. */
+  override def getNextBatch(start: Option[Offset]): Option[Batch] = {
+    val beginOffset: KafkaSourceOffset = KafkaSourceOffset.from(start).getOrElse(initialOffsets)
+    val latestOffset = getLatestOffsets()
+    logDebug(s"Latest offset: $latestOffset")
+
+    val offsetRanges = beginOffset to latestOffset
+    val kafkaParams = params
+    val encodingFunc = encoder.toRow _
+    val sparkContext = sqlContext.sparkContext
+
+    if (offsetRanges.nonEmpty) {
+      val rdd = KafkaUtils.createRDD[Array[Byte], Array[Byte], DefaultDecoder, DefaultDecoder](
+        sparkContext, kafkaParams, offsetRanges.toArray)
+      logInfo("Creating DF with offset ranges: " + offsetRanges)
+      Some(new Batch(latestOffset, sqlContext.createDataset(rdd).toDF))
+    } else {
+      None
+    }
+  }
+
+  def toDS(): Dataset[(Array[Byte], Array[Byte])] = {
+    toDF.as[(Array[Byte], Array[Byte])]
+  }
+
+  def toDF(): DataFrame = {
+    new DataFrame(sqlContext, logicalPlan)
+  }
+
+  /** Get latest offsets from Kafka. */
+  private def getLatestOffsets(): KafkaSourceOffset = {
+    val partitionLeaders = KafkaCluster.checkErrors(kc.findLeaders(topicAndPartitions))
+    val leadersAndOffsets = KafkaCluster.checkErrors(kc.getLatestLeaderOffsets(topicAndPartitions))
+    KafkaSourceOffset(leadersAndOffsets.map { x => (x._1, x._2.offset) })
+  }
+
+  /** Get the initial offsets from Kafka for the source to start from. */
+  private def getInitialOffsets(): KafkaSourceOffset = {
+    if (params.get("auto.offset.reset").map(_.toLowerCase) == Some("smallest")) {
```

Comment:
typically we do all extraction / validation of the options in the provider and then the arguments to the source are typed (instead of passing the map into the class itself).  The nice thing about this is it centralizes the location for all configuration.